# Kids First Data Resource Center Pacific Biosciences Long Reads Alignment and Variant Calling Workflow

<p align="center">
  <img src="https://github.com/d3b-center/d3b-research-workflows/raw/master/doc/kfdrc-logo-sm.png">
</p>

The Kids First Data Resource Center (KFDRC) Pacific Biosciences (PacBio)
Long Reads Alignment and Variant Calling Workflow is a Common Workflow Language
(CWL) implementation of various softwares used to take reads information
generated by PacBio long reads sequencers and generate alignment and variant
information. This pipeline was made possible thanks to significant software and
support contributions from both Sentieon and Wang Genomics Lab. For more
information on our collaborators, check out their websites:
- Sentieon: https://www.sentieon.com/
- Wang Genomics Lab: https://wglab.org/

## Relevant Softwares and Versions
- [samtools head](http://www.htslib.org/doc/samtools-head.html): `1.17`
- [samtools fastq](http://www.htslib.org/doc/samtools-fastq.html): `1.15.1`
- [Sentieon Minimap2](https://support.sentieon.com/manual/usages/general/?highlight=minimap2#minimap2-binary): `202112.01`
- [Sentieon util sort](https://support.sentieon.com/manual/usages/general/?highlight=minimap2#util-binary): `202112.01`
- [Sentieon DNAScope HiFi](https://support.sentieon.com/manual/): `202112.01`
- [Sentieon LongReadSV](https://support.sentieon.com/manual/): `202112.06`
- [LongReadSum](https://github.com/WGLab/LongReadSum#readme): `1.2.0`
- [Sniffles](https://github.com/fritzsedlazeck/Sniffles#readme): `2.0.7`
- [pbsv](https://github.com/PacificBiosciences/pbsv#readme): `2.9.0`

## Input Files
- `input_unaligned_bam`: The primary input of the PacBio Long Reads Workflow is an unaligned BAM and associated index.
- `indexed_reference_fasta`: Any suitable human reference genome. KFDRC uses `Homo_sapiens_assembly38.fasta` from Broad Institute.

## Output Files
- `dnascope_small_variants`: BGZIP and TABIX indexed VCF containing small variant calls made by Sentieon DNAScope HiFi on `minimap2_aligned_bam`.
- `longreadsum_bam_metrics`: BGZIP TAR containing various metrics collected by LongReadSum from the `minimap2_aligned_bam`.
- `minimap2_aligned_bam`: Indexed BAM file containing reads from the `input_unaligned_bam` aligned to the `indexed_reference_fasta`.
- `pbsv_structural_variants`: BGZIP and TABIX indexed VCF containing structural variant calls made by pbsv on the `minimap2_aligned_bam`.
- `sniffles_structural_variants`: BGZIP and TABIX indexed VCF containing structural variant calls made by Sniffles on the `minimap2_aligned_bam`.
- `longreadsv_structural_variants`: BGZIP and TABIX indexed VCF containing structural variant calls made by Sentieon LongReadSV on the `minimap2_aligned_bam`.

## Generalized Process
1. Read group information (`@RG`) is harvested from the `input_unaligned_bam` header using `samtools head` and `grep`.
1. If user provides `biospecimen_name` input, that value replaces the `SM` value pulled in the preceeding step.
1. Align `input_unaligned_bam` to `indexed_reference_fasta` with tohe above `@RG` information using samtools fastq, Sentieon Minimap2, and Sentieon sort.
1. Generate long reads alignment metrics from the `minimap2_aligned_bam` using LongReadSum.
1. Generate structural variant calls from the `minimap2_aligned_bam` using pbsv.
1. Generate structural variant calls from the `minimap2_aligned_bam` using Sniffles.
1. Generate structural variant calls from the `minimap2_aligned_bam` using Sentieon LongReadSV.
1. If the reads are not CLR, Generate small variant from the `minimap2_aligned_bam` using Sentieon DNAScope HiFi.

## Basic Info
- [D3b dockerfiles](https://github.com/d3b-center/bixtools)
- Testing Tools:
    - [Seven Bridges Cavatica Platform](https://cavatica.sbgenomics.com/)
    - [Common Workflow Language reference implementation (cwltool)](https://github.com/common-workflow-language/cwltool/)

## References
- KFDRC AWS s3 bucket: s3://kids-first-seq-data/broad-references/
- Cavatica: https://cavatica.sbgenomics.com/u/kfdrc-harmonization/kf-references/
- Broad Institute Goolge Cloud: https://console.cloud.google.com/storage/browser/genomics-public-data/resources/broad/hg38/v0/
